File: lib/ask-markup.ts
import { Prisma, type PrismaClient } from "@prisma/client";
import { pipSize } from "@/lib/group-pricing";
import { resolvePricingV2, resolveEffectiveSpreadMarkup, type ResolvedSpread } from "@/lib/pricing-engine";

// Owner decision (2026-09-26): EVERY ask-side execution uses the account's marked-up ask -- a BUY open (as before)
// AND a SELL close, including a SELL's SL / TP trigger, its stop-out, and every valuation of an open SELL at the price
// it would close at (floating P/L, equity, used margin, margin level, the displayed current price). A BUY closes at
// the raw bid, a SELL opens at the raw bid: the bid is never marked up.
//
// This module is the one definition of "this account's ask for this symbol". The rule is resolved with EXACTLY the
// fill path's precedence (lib/pricing-engine.ts resolveFillPricing):
//   - pricing engine off: GroupSymbolConfig.spreadMarkup, else BrokerSymbol.spreadMarkup (resolveSymbolPricing);
//   - pricing engine on: AccountSymbolConfig > AccountTypeSymbolConfig > AccountType > GroupSymbolConfig > BrokerSymbol,
//     target-total-spread mode included (resolvePricingV2 + resolveEffectiveSpreadMarkup);
//   - the broker's coverage account (a COVERAGE group) is always raw: its legs mirror the real market, never the
//     broker's own markup (lib/coverage.ts).
// The price itself is applySpreadMarkup's formula (ask + markup x pip), so a SELL closing and a BUY opening on the same
// tick at the same account pay the same ask. docs/contracts/ask-markup-vectors.json pins it for the engine (Rust) and
// the clients (C#, WebTrader); lib/ask-markup.test.ts keeps that file honest.

export type AskRule =
  | { mode: "markup"; markupPips: Prisma.Decimal; digits: number }
  | { mode: "target"; targetPips: Prisma.Decimal; fallbackPips: Prisma.Decimal | null; digits: number };

export const RAW_ASK = (digits: number): AskRule => ({ mode: "markup", markupPips: new Prisma.Decimal(0), digits });

/** The rule from a resolved spread (pricing engine on) or a plain markup (off). */
export function askRuleFromSpread(spread: ResolvedSpread, digits: number): AskRule {
  return spread.mode === "target"
    ? { mode: "target", targetPips: spread.targetTotalSpreadPips, fallbackPips: spread.fallbackSpreadMarkup, digits }
    : { mode: "markup", markupPips: spread.spreadMarkup, digits };
}

/** The markup in pips this rule applies at this tick (target mode: whatever lifts the live spread to the target, floored
 *  at 0; no live spread -> the level's own fallback markup, else 0) -- resolveEffectiveSpreadMarkup, unchanged. */
export function markupPipsAt(rule: AskRule, bid: Prisma.Decimal.Value, ask: Prisma.Decimal.Value): Prisma.Decimal {
  if (rule.mode === "markup") return rule.markupPips;
  const base = new Prisma.Decimal(ask).sub(new Prisma.Decimal(bid)).div(pipSize(rule.digits));
  return resolveEffectiveSpreadMarkup({ mode: "target", targetTotalSpreadPips: rule.targetPips, fallbackSpreadMarkup: rule.fallbackPips }, base).markup;
}

/** The account's ask: raw ask + markup x pip (lib/group-pricing.ts applySpreadMarkup's formula). */
export function markedUpAsk(rule: AskRule, bid: Prisma.Decimal.Value, ask: Prisma.Decimal.Value): Prisma.Decimal {
  const raw = new Prisma.Decimal(ask);
  const m = markupPipsAt(rule, bid, ask);
  return m.isZero() ? raw : raw.add(m.mul(pipSize(rule.digits)));
}

/** The price an open position closes at for THIS account: a BUY at the raw bid, a SELL at the account's ask. */
export function accountClosePrice(side: "BUY" | "SELL", bid: Prisma.Decimal.Value, ask: Prisma.Decimal.Value, rule: AskRule | null | undefined): Prisma.Decimal {
  if (side === "BUY") return new Prisma.Decimal(bid);
  return rule ? markedUpAsk(rule, bid, ask) : new Prisma.Decimal(ask);
}

/** The wire form a client needs to apply the same rule on every tick (lib/trade-api.ts spreadRuleFromPrice): the
 *  markup in price units at this tick, and the target in price units when in target mode. */
export function askRuleWire(rule: AskRule, bid?: Prisma.Decimal.Value | null, ask?: Prisma.Decimal.Value | null): { askMarkup: string; spreadRule: "markup" | "target"; targetSpread?: string } {
  const pip = pipSize(rule.digits);
  if (rule.mode === "target") {
    const m = bid != null && ask != null ? markupPipsAt(rule, bid, ask) : (rule.fallbackPips ?? new Prisma.Decimal(0));
    return { askMarkup: m.mul(pip).toString(), spreadRule: "target", targetSpread: rule.targetPips.mul(pip).toString() };
  }
  return { askMarkup: rule.markupPips.mul(pip).toString(), spreadRule: "markup" };
}

type Db = PrismaClient | Prisma.TransactionClient;
export type AskRuleKey = { accountId: string; symbolId: string };
export type AskRules = { get(accountId: string, symbolId: string): AskRule | null };

/**
 * The ask rule of every (account, symbol) pair asked for, in a constant number of queries whatever the number of pairs
 * (accounts, brokers, broker symbols, symbols, group configs, and -- for a pricing-engine broker -- account types and
 * their / the accounts' own per-symbol configs). A pair whose account or broker symbol is gone resolves to null, which
 * every caller treats as the raw ask.
 */
export async function loadAskRules(db: Db, pairs: AskRuleKey[]): Promise<AskRules> {
  const map = new Map<string, AskRule>();
  const key = (a: string, s: string) => `${a}|${s}`;
  const result: AskRules = { get: (a, s) => map.get(key(a, s)) ?? null };
  if (pairs.length === 0) return result;
  const accountIds = [...new Set(pairs.map((p) => p.accountId))];
  const symbolIds = [...new Set(pairs.map((p) => p.symbolId))];

  const [accounts, symbols] = await Promise.all([
    db.account.findMany({ where: { id: { in: accountIds } }, select: { id: true, brokerId: true, groupId: true, accountTypeId: true, group: { select: { category: true } } } }),
    db.symbol.findMany({ where: { id: { in: symbolIds } }, select: { id: true, digits: true } }),
  ]);
  const brokerIds = [...new Set(accounts.map((a) => a.brokerId))];
  const groupIds = [...new Set(accounts.map((a) => a.groupId))];
  const [brokers, brokerSymbols, groupConfigs] = await Promise.all([
    db.broker.findMany({ where: { id: { in: brokerIds } }, select: { id: true, pricingEngineEnabled: true, coverageAccountId: true } }),
    db.brokerSymbol.findMany({ where: { brokerId: { in: brokerIds }, symbolId: { in: symbolIds } }, select: { brokerId: true, symbolId: true, spreadMarkup: true, commissionPerLot: true, swapLong: true, swapShort: true } }),
    db.groupSymbolConfig.findMany({ where: { groupId: { in: groupIds }, symbolId: { in: symbolIds } }, select: { groupId: true, symbolId: true, spreadMarkup: true, targetTotalSpreadPips: true, commissionPerLot: true, swapLong: true, swapShort: true } }),
  ]);
  const brokerById = new Map(brokers.map((b) => [b.id, b]));
  const engineAccounts = accounts.filter((a) => brokerById.get(a.brokerId)?.pricingEngineEnabled);
  const typeIds = [...new Set(engineAccounts.map((a) => a.accountTypeId).filter((t): t is string => t != null))];
  const [types, typeConfigs, accountConfigs] = engineAccounts.length === 0
    ? [[], [], []]
    : await Promise.all([
        typeIds.length ? db.accountType.findMany({ where: { id: { in: typeIds } }, select: { id: true, spreadMarkup: true, commissionPerLot: true, swapLong: true, swapShort: true, swapFree: true } }) : Promise.resolve([]),
        typeIds.length ? db.accountTypeSymbolConfig.findMany({ where: { accountTypeId: { in: typeIds }, symbolId: { in: symbolIds } }, select: { accountTypeId: true, symbolId: true, spreadMarkup: true, targetTotalSpreadPips: true, commissionPerLot: true, swapLong: true, swapShort: true } }) : Promise.resolve([]),
        db.accountSymbolConfig.findMany({ where: { accountId: { in: engineAccounts.map((a) => a.id) }, symbolId: { in: symbolIds } }, select: { accountId: true, symbolId: true, spreadMarkup: true, targetTotalSpreadPips: true, commissionPerLot: true, swapLong: true, swapShort: true } }),
      ]);

  const accountById = new Map(accounts.map((a) => [a.id, a]));
  const digitsBySymbol = new Map(symbols.map((s) => [s.id, s.digits]));
  const brokerSymbolBy = new Map(brokerSymbols.map((b) => [`${b.brokerId}|${b.symbolId}`, b]));
  const groupConfigBy = new Map(groupConfigs.map((c) => [`${c.groupId}|${c.symbolId}`, c]));
  const typeById = new Map(types.map((t) => [t.id, t]));
  const typeConfigBy = new Map(typeConfigs.map((c) => [`${c.accountTypeId}|${c.symbolId}`, c]));
  const accountConfigBy = new Map(accountConfigs.map((c) => [`${c.accountId}|${c.symbolId}`, c]));

  for (const { accountId, symbolId } of pairs) {
    const a = accountById.get(accountId);
    const digits = digitsBySymbol.get(symbolId);
    if (!a || digits == null) continue;
    const broker = brokerById.get(a.brokerId);
    const bs = brokerSymbolBy.get(`${a.brokerId}|${symbolId}`);
    if (!broker || !bs) continue;
    if (a.group?.category === "COVERAGE" || broker.coverageAccountId === a.id) {
      map.set(key(accountId, symbolId), RAW_ASK(digits));
      continue;
    }
    const gsc = groupConfigBy.get(`${a.groupId}|${symbolId}`) ?? null;
    if (!broker.pricingEngineEnabled) {
      // resolveSymbolPricing: the group's markup when it sets one, else the broker's
      map.set(key(accountId, symbolId), { mode: "markup", markupPips: gsc?.spreadMarkup ?? bs.spreadMarkup, digits });
      continue;
    }
    const resolved = resolvePricingV2({
      accountSymbolConfig: accountConfigBy.get(`${accountId}|${symbolId}`) ?? null,
      accountTypeSymbolConfig: a.accountTypeId ? (typeConfigBy.get(`${a.accountTypeId}|${symbolId}`) ?? null) : null,
      accountType: a.accountTypeId ? (typeById.get(a.accountTypeId) ?? null) : null,
      groupSymbolConfig: gsc,
      brokerSpreadMarkup: bs.spreadMarkup,
      brokerCommissionPerLot: bs.commissionPerLot,
      brokerSwapLong: bs.swapLong,
      brokerSwapShort: bs.swapShort,
      accountSwapFree: null,
      groupSwapFree: null,
    });
    map.set(key(accountId, symbolId), askRuleFromSpread(resolved.spread, digits));
  }
  return result;
}

/** One account's rules for every symbol it is asked about (the common single-account case). */
export async function loadAccountAskRules(db: Db, accountId: string, symbolIds: string[]): Promise<(symbolId: string) => AskRule | null> {
  const rules = await loadAskRules(db, symbolIds.map((symbolId) => ({ accountId, symbolId })));
  return (symbolId) => rules.get(accountId, symbolId);
}

/** The rules a set of positions needs to be valued at their close-side price: only a SELL closes at the ask, so only
 *  SELL positions are resolved (none = no query at all). */
export function loadSellAskRules(db: Db, positions: { accountId: string; symbolId: string; side: "BUY" | "SELL" }[]): Promise<AskRules> {
  return loadAskRules(db, positions.filter((p) => p.side === "SELL").map((p) => ({ accountId: p.accountId, symbolId: p.symbolId })));
}

/** A position's close-side ask for valuation: the account's ask for a SELL; a BUY never reads it (it closes at bid). */
export function valuationAsk(rules: AskRules, p: { accountId: string; symbolId: string; side: "BUY" | "SELL" }, bid: Prisma.Decimal.Value, ask: Prisma.Decimal.Value): Prisma.Decimal {
  return p.side === "SELL" ? accountClosePrice("SELL", bid, ask, rules.get(p.accountId, p.symbolId)) : new Prisma.Decimal(ask);
}

File: engine/market-data/src/ask_markup.rs
//! The account's ask (owner decision 2026-09-26): every ask-side execution uses the account's marked-up ask -- a BUY
//! open AND a SELL close, including a SELL's SL / TP trigger, its stop-out, and every valuation of an open SELL at the
//! price it would close at. A BUY closes at the raw bid; the bid is never marked up.
//!
//! The Rust port of lib/ask-markup.ts, pinned to it by docs/contracts/ask-markup-vectors.json (the test below reads the
//! file). Lives here, not in order-management, because market-data's risk hook needs it too (SL / TP levels and resting
//! BUY entries are checked against the account's ask) and cannot depend on order-management.
//!
//! Resolution, per (account, symbol), exactly the web's fill path (lib/pricing-engine.ts resolveFillPricing):
//! - no BrokerSymbol row: no rule (the raw ask), as the web's loadAskRules skips the pair;
//! - the broker's coverage account (its group is COVERAGE, or Broker.coverageAccountId): raw, always;
//! - pricing engine off: GroupSymbolConfig.spreadMarkup, else BrokerSymbol.spreadMarkup (a target is never read);
//! - pricing engine on: the first level that sets a spread (markup or target; target wins within a level, its markup
//!   rides along as the fallback) of AccountSymbolConfig, AccountTypeSymbolConfig, then AccountType.spreadMarkup (flat,
//!   markup only), then GroupSymbolConfig, then BrokerSymbol.spreadMarkup.
//!
//! Price: pip = 10^-(digits-1) (digits <= 1: 1). markup mode: ask + markup x pip. target mode: markup = target - the live
//! raw spread in pips, floored at 0 -> ask + markup x pip.

use protocol::OrderSide;
use rust_decimal::Decimal;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AskRule {
    Markup { markup_pips: Decimal, digits: i32 },
    Target { target_pips: Decimal, fallback_pips: Option<Decimal>, digits: i32 },
}

/// 1 pip in price units (lib/group-pricing.ts pipSize, pricing.rs pip_size).
pub fn pip_size(digits: i32) -> Decimal {
    Decimal::new(1, digits.saturating_sub(1).max(0) as u32)
}

/// One per-symbol config level: a markup and / or a target, in pips.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct Level {
    pub markup: Option<Decimal>,
    pub target: Option<Decimal>,
}

/// Everything the rule of one (account, symbol) depends on, as read from the book.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct AskLevels {
    pub pricing_engine: bool,
    pub coverage: bool,
    pub digits: i32,
    /// BrokerSymbol.spreadMarkup; None = no BrokerSymbol row for this broker and symbol.
    pub broker: Option<Decimal>,
    pub group: Level,
    /// AccountType.spreadMarkup (the flat level: markup only).
    pub account_type: Option<Decimal>,
    pub account_type_symbol: Level,
    pub account_symbol: Level,
}

fn level_rule(l: &Level, digits: i32) -> Option<AskRule> {
    if let Some(target_pips) = l.target {
        return Some(AskRule::Target { target_pips, fallback_pips: l.markup, digits });
    }
    l.markup.map(|markup_pips| AskRule::Markup { markup_pips, digits })
}

/// The rule, or None (the raw ask) when the broker does not carry the symbol.
pub fn resolve(l: &AskLevels) -> Option<AskRule> {
    let broker = l.broker?;
    let digits = l.digits;
    if l.coverage {
        return Some(AskRule::Markup { markup_pips: Decimal::ZERO, digits });
    }
    if !l.pricing_engine {
        return Some(AskRule::Markup { markup_pips: l.group.markup.unwrap_or(broker), digits });
    }
    Some(
        level_rule(&l.account_symbol, digits)
            .or_else(|| level_rule(&l.account_type_symbol, digits))
            .or_else(|| l.account_type.map(|markup_pips| AskRule::Markup { markup_pips, digits }))
            .or_else(|| level_rule(&l.group, digits))
            .unwrap_or(AskRule::Markup { markup_pips: broker, digits }),
    )
}

/// The markup in pips this rule applies at this tick (lib/pricing-engine.ts resolveEffectiveSpreadMarkup).
pub fn markup_pips_at(rule: &AskRule, bid: Decimal, ask: Decimal) -> Decimal {
    match *rule {
        AskRule::Markup { markup_pips, .. } => markup_pips,
        AskRule::Target { target_pips, digits, .. } => {
            let base = (ask - bid) / pip_size(digits);
            let diff = target_pips - base;
            if diff.is_sign_negative() && !diff.is_zero() { Decimal::ZERO } else { diff }
        }
    }
}

/// The account's ask: raw ask + markup x pip (the BUY fill price, pricing.rs apply_spread_markup's formula).
pub fn account_ask(rule: &AskRule, bid: Decimal, ask: Decimal) -> Decimal {
    let m = markup_pips_at(rule, bid, ask);
    if m.is_zero() {
        return ask;
    }
    let digits = match *rule {
        AskRule::Markup { digits, .. } | AskRule::Target { digits, .. } => digits,
    };
    ask + m * pip_size(digits)
}

/// The price an open position closes at for its account: a BUY at the raw bid, a SELL at the account's ask (no rule =
/// the raw ask).
pub fn close_price(side: OrderSide, bid: Decimal, ask: Decimal, rule: Option<&AskRule>) -> Decimal {
    match side {
        OrderSide::Buy => bid,
        OrderSide::Sell => rule.map_or(ask, |r| account_ask(r, bid, ask)),
    }
}

/// The joins a query adds to read the levels of its rows' (account, symbol). The query must alias the account `a` and
/// the symbol `s`; every joined table is prefixed `am_` so it never clashes with the query's own aliases.
pub const LEVELS_JOINS: &str = r#"
    LEFT JOIN "Broker" am_b ON am_b.id = a."brokerId"
    LEFT JOIN "Group" am_g ON am_g.id = a."groupId"
    LEFT JOIN "BrokerSymbol" am_bs ON am_bs."brokerId" = a."brokerId" AND am_bs."symbolId" = s.id
    LEFT JOIN "GroupSymbolConfig" am_gsc ON am_gsc."groupId" = a."groupId" AND am_gsc."symbolId" = s.id
    LEFT JOIN "AccountType" am_at ON am_at.id = a."accountTypeId"
    LEFT JOIN "AccountTypeSymbolConfig" am_atsc ON am_atsc."accountTypeId" = a."accountTypeId" AND am_atsc."symbolId" = s.id
    LEFT JOIN "AccountSymbolConfig" am_asc ON am_asc."accountId" = a.id AND am_asc."symbolId" = s.id "#;

/// The columns those joins provide, read back by `levels_from_row`.
pub const LEVELS_COLUMNS: &str = r#"
    COALESCE(am_b."pricingEngineEnabled", false) AS am_engine,
    (COALESCE(am_g.category::text = 'COVERAGE', false) OR COALESCE(am_b."coverageAccountId" = a.id, false)) AS am_coverage,
    s.digits AS am_digits,
    am_bs."spreadMarkup" AS am_broker,
    am_gsc."spreadMarkup" AS am_g_m, am_gsc."targetTotalSpreadPips" AS am_g_t,
    am_at."spreadMarkup" AS am_at_m,
    am_atsc."spreadMarkup" AS am_ats_m, am_atsc."targetTotalSpreadPips" AS am_ats_t,
    am_asc."spreadMarkup" AS am_as_m, am_asc."targetTotalSpreadPips" AS am_as_t "#;

/// The levels of one row of a query that selected `LEVELS_COLUMNS` (with `LEVELS_JOINS`).
pub fn levels_from_row(row: &sqlx::postgres::PgRow) -> Result<AskLevels, sqlx::Error> {
    use sqlx::Row;
    let lvl = |m: &str, t: &str| -> Result<Level, sqlx::Error> { Ok(Level { markup: row.try_get(m)?, target: row.try_get(t)? }) };
    Ok(AskLevels {
        pricing_engine: row.try_get("am_engine")?,
        coverage: row.try_get("am_coverage")?,
        digits: row.try_get("am_digits")?,
        broker: row.try_get("am_broker")?,
        group: lvl("am_g_m", "am_g_t")?,
        account_type: row.try_get("am_at_m")?,
        account_type_symbol: lvl("am_ats_m", "am_ats_t")?,
        account_symbol: lvl("am_as_m", "am_as_t")?,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use rust_decimal_macros::dec;
    use std::str::FromStr;

    fn vectors() -> serde_json::Value {
        let path = concat!(env!("CARGO_MANIFEST_DIR"), "/../../docs/contracts/ask-markup-vectors.json");
        serde_json::from_str(&std::fs::read_to_string(path).expect("ask-markup-vectors.json")).unwrap()
    }
    fn dec_opt(v: &serde_json::Value) -> Option<Decimal> {
        v.as_str().map(|s| Decimal::from_str(s).unwrap())
    }
    fn level(v: &serde_json::Value) -> Level {
        if v.is_null() {
            return Level::default();
        }
        Level { markup: dec_opt(&v["spreadMarkup"]), target: dec_opt(&v["targetTotalSpreadPips"]) }
    }
    /// The web writes decimals normalised (decimal.js toString: no trailing zeros).
    fn s(d: Decimal) -> String {
        d.normalize().to_string()
    }

    #[test]
    fn every_resolution_case_of_the_vectors() {
        let v = vectors();
        let cases = v["resolution"].as_array().unwrap();
        assert!(cases.len() >= 14);
        for c in cases {
            let name = c["name"].as_str().unwrap();
            let l = AskLevels {
                pricing_engine: c["pricingEngineEnabled"].as_bool().unwrap(),
                coverage: c["coverage"].as_bool().unwrap(),
                digits: c["digits"].as_i64().unwrap() as i32,
                broker: dec_opt(&c["broker"]),
                group: level(&c["group"]),
                account_type: if c["accountType"].is_null() { None } else { dec_opt(&c["accountType"]["spreadMarkup"]) },
                account_type_symbol: level(&c["accountTypeSymbol"]),
                account_symbol: level(&c["accountSymbol"]),
            };
            let got = resolve(&l).unwrap();
            let e = &c["expected"];
            match got {
                AskRule::Markup { markup_pips, .. } => {
                    assert_eq!(e["mode"], "markup", "{name}");
                    assert_eq!(s(markup_pips), e["markupPips"].as_str().unwrap(), "{name}");
                }
                AskRule::Target { target_pips, fallback_pips, .. } => {
                    assert_eq!(e["mode"], "target", "{name}");
                    assert_eq!(s(target_pips), e["targetPips"].as_str().unwrap(), "{name}");
                    assert_eq!(fallback_pips.map(s), e["fallbackPips"].as_str().map(str::to_string), "{name}");
                }
            }
        }
    }

    #[test]
    fn every_price_case_of_the_vectors() {
        let v = vectors();
        let cases = v["prices"].as_array().unwrap();
        assert!(cases.len() >= 9);
        for c in cases {
            let name = c["name"].as_str().unwrap();
            let digits = c["digits"].as_i64().unwrap() as i32;
            let r = &c["rule"];
            let rule = if r["mode"] == "markup" {
                AskRule::Markup { markup_pips: dec_opt(&r["markupPips"]).unwrap(), digits }
            } else {
                AskRule::Target { target_pips: dec_opt(&r["targetPips"]).unwrap(), fallback_pips: dec_opt(&r["fallbackPips"]), digits }
            };
            let (bid, ask) = (dec_opt(&c["bid"]).unwrap(), dec_opt(&c["ask"]).unwrap());
            let e = &c["expected"];
            assert_eq!(s(markup_pips_at(&rule, bid, ask)), e["markupPips"].as_str().unwrap(), "{name}: markupPips");
            assert_eq!(s(account_ask(&rule, bid, ask)), e["accountAsk"].as_str().unwrap(), "{name}: accountAsk");
            assert_eq!(s(close_price(OrderSide::Buy, bid, ask, Some(&rule))), e["buyClose"].as_str().unwrap(), "{name}: buyClose");
            assert_eq!(s(close_price(OrderSide::Sell, bid, ask, Some(&rule))), e["sellClose"].as_str().unwrap(), "{name}: sellClose");
        }
    }

    #[test]
    fn no_broker_symbol_means_no_rule_and_no_rule_means_the_raw_ask() {
        assert_eq!(resolve(&AskLevels { broker: None, pricing_engine: true, group: Level { markup: Some(dec!(9)), target: None }, ..Default::default() }), None);
        assert_eq!(close_price(OrderSide::Sell, dec!(1), dec!(1.5), None), dec!(1.5));
    }
}
